### PR TITLE
Make `pytest.warns()` re-emit all non-matching warnings

### PR DESCRIPTION
**Pull request by reaganjlee**
PR Description: 

This issue fixes part of issue #9288 by allowing ``pytest.warns()`` to emit all unmatched warnings. 

One question I have with the design is how to handle the exceptions once we re-emit them. Having ``pytest.warns()`` calls return all exceptions will require refactoring of other tests. There doesn't seem to be a way to check whether the code is within a WarningsChecker frame, so a way to solve this is an optional-argument for re-emitting?

### CODE REVIEW

File: src/_pytest/recwarn.py
Comment (RonnyPfannschmidt) [inline src/_pytest/recwarn.py:326]: the logic here is slightly terrifying at the nesting level, at first glance i had the impression this must be wrong

however was we break right after it its indeed correct

**Comment by reaganjlee** [comment]
Since we want to reduce number of refactors (and potentially, number of optional arguments), we can assume that if a user specifies the particular class they'd like to catch warnings from, then they would use this same warning type for similar checks, leading us to filter that exception class and remitting the different, unexpected exceptions from the one specified? If there is a specific match given, then ignore this and do the usual re-emit besides the one/s found. 

This may or may not still lead to some of the existing, unrelated tests failing, though I can implement this soon and see how that goes

**Comment by RonnyPfannschmidt** [comment]
I'm currently not able to allocate some time to write the details down to reiterate the options

My current recommendation would be start with a implementation that filters all matching warnings and then Iterating on documentation for a future api where the other use cases can be served

**Comment by reaganjlee** [comment]
Let me know if there are any documentation formatting to fix for this

**Comment by Zac-HD** [comment]
Hey Reagan - let's try to get this merged soon!

- Rebase on the latest `main` branch
- I think that we should re-emit all and only warnings which do not match (by type and possibly regex pattern).
- We'll need to fix the `mypy` warnings from pre-commit ci; add a changelog entry, and make sure you're in the authors list

and I think that's it; your docs look good to me and I don't think we need to handle any additional complexity here.

File: src/_pytest/recwarn.py
Comment (Zac-HD) [inline src/_pytest/recwarn.py:155]:
```suggestion
        Unlike the stdlib :func:`warnings.catch_warnings` context manager,
        unmatched warnings will be re-emitted when the context closes.
```

File: src/_pytest/recwarn.py
Comment (Zac-HD) [inline src/_pytest/recwarn.py:111]: I think this new feature should be unconditionally enabled, which will simplify it for both users and maintainers.

File: src/_pytest/recwarn.py
Comment (Zac-HD) [inline src/_pytest/recwarn.py:307]: I think that this logic is wrong: when we capture multiple warnings which match the type and message, we should remove _all_ matching warnings, not just the first.  We'll also need to add a test for this.

File: src/_pytest/recwarn.py
Comment (Zac-HD) [inline src/_pytest/recwarn.py:330]: We might want to extract this to a new method `def matches(self, warning) -> bool:` to share the logic between here and the re-emit function.

**Review by Zac-HD — COMMENTED** [review]
As you note, we'll need a changelog, docs, and adding you to the authors list (🎊); after those, fixing the mypy CI errors, and comments below I think we'll be ready to merge!

File: src/_pytest/recwarn.py
Comment (The-Compiler) [inline src/_pytest/recwarn.py:344]: Even if those methods are only used from inside `__exit__`, it seems very odd to me to define an instance method as a function inside here and pass `self` to it - why not make them normal methods of `WarningChecker` instead?

Comment (reaganjlee) [reply src/_pytest/recwarn.py]: Just used them to solve arg and mypy errors -- will look into mypy without the instance methods

File: src/_pytest/recwarn.py
Comment (nicoddemus) [inline src/_pytest/recwarn.py:154]:
```suggestion
        Since pytest 8.0, unlike the stdlib :func:`warnings.catch_warnings` context manager,
```

**Review by nicoddemus — COMMENTED** [review]
Thanks @reaganjlee and everyone for the comments.

Left a few smaller comments too, other than that LGTM. :+1:

**Review by The-Compiler — COMMENTED** [review]
I don't think I'll be able to review this - I barely ever use warnings (given my main project is an application rather than a library), so I'd rather leave this to someone else.

**Comment by Zac-HD** [comment]
Thanks @nicoddemus, all done!

Since [Python 3.7 is now end-of-life](https://devguide.python.org/versions/), we should probably test on a newer version of PyPy (3.8 through 3.10 are now supported) and remove the old CI jobs.  I'd rather not add that into this long-running PR though 😅

File: src/_pytest/recwarn.py
Comment (nicoddemus) [inline src/_pytest/recwarn.py:303]: Not sure what this TODO means here... can we clarify it? Also not clear if this is something meant to be done still in this PR, or later.

**Comment by nicoddemus** [comment]
> we should probably test on a newer version of PyPy (3.8 through 3.10 are now supported) and remove the old CI jobs. I'd rather not add that into this long-running PR though

Agreed, we can do any adjustments later after `main` is adapted. :+1:

File: src/_pytest/recwarn.py
Comment (Zac-HD) [reply src/_pytest/recwarn.py:303]: Ah, that's a note-to-self about rebasing https://github.com/pytest-dev/pytest/pull/11129 which I'll do immediately after this is merged.

**Review by Zac-HD — APPROVED** [review]

**Comment by pllim** [comment]
I *think* this broke downstream tests reported at https://github.com/astropy/astropy/issues/15015 . Our intention of the affected tests are to make sure a particular warning (that we are checking with `pytest.warns`) is emitted, not necessarily that it has to be the only warning. If we want to make sure that it is the only warning, we check the `len` later. I suspect this change would break multiple downstream packages when released. Is there a way to get the old behavior back? Or did I misdiagnose the downstream failures?

**Comment by Zac-HD** [comment]
I think you're correct in tracing the change to this PR.

We're aware that this will lead to more warnings being emitted, which can make tests fail with `-Werror` (that's why [it's documented as a breaking change](https://github.com/pytest-dev/pytest/pull/10937/files#diff-9393c17d4888c396708f765db922e60cfe8c2dcd03bd1602fd9b2569dd72ce79)!), but we think ceasing to hide unrelated warnings is worth making that change in Pytest 8.0 - see https://github.com/pytest-dev/pytest/issues/11126 for the release plan.

To ignore any other warnings, without asserting that there must be other warnings, you can use the stdlib `warnings.catch_warnings()` + `warnings.simplefilter("ignore")` functions [as we document here](https://docs.pytest.org/en/latest/how-to/capture-warnings.html#additional-use-cases-of-warnings-in-tests).  I contributed a bit of API sugar so [from Python 3.11 you can use](https://docs.python.org/3/library/warnings.html#warnings.catch_warnings) `with catch_warnings(action="ignore"):`, but it might be a while before you can rely on that everywhere.

----------

I did notice skimming through the failures that [this test](https://github.com/astropy/astropy/actions/runs/5441155135/jobs/9894791695#step:7:1985) ([source here](https://github.com/astropy/astropy/blob/645e587f329342bb6d748e1d3398bd3eadab71ab/astropy/io/fits/tests/test_hdulist.py#L1136)) could be affected by https://github.com/pytest-dev/pytest/pull/11129 - up until now, a `warns()` within a `raises()` did literally nothing at all 😓.  [This gzip warning](https://github.com/astropy/astropy/actions/runs/5441155135/jobs/9894791695#step:7:2227) also looks real, though I don't know enough about FITS to tell for most of the rest.

**Comment by pllim** [comment]
Thanks for the clarification! I guess I'll have to go in and clean stuff up. 😸

**Pull request merged**
